Flex Stacker in API docs and reference (#19176)
OpenTrons/opentrons-api: api/docs/v2/modules/stacker.rst

:og:description: How to use the Flex Stacker Module in a Python protocol. 

.. _stacker:

*******************
Flex Stacker Module
*******************

The Flex Stacker is an external module that provides automated storage and dispensing of Flex tip racks, well plates, or reservoirs. The Stacker's attached shuttle moves stored labware from the stack to add to the Flex deck during a protocol. 

The Stacker is represented in code by a :py:class:`.FlexStackerContext` object that includes methods for retrieving and storing labware. You can also use helper commands in your protocol to calculate how many labware the Stacker can store at once. 

Loading and Deck Slots
========================

> Up to four Stacker Modules can be attached to the right side of your Flex, creating deck slots A4–D4. Each Stacker occupies a deck slot in column 4, with the shuttle in column 3. 

Reviewer: This is off by one. Shuttle is in A4 (staging area). Stacker has a regular deck slot (or waste chute) in column 3.

Start by loading each Stacker as you would any other module: 

.. code-block:: python

   stacker_1 = protocol.load_module(
       module_name="flexStackerModuleV1",
       location="A4"
   )
   stacker_2 = protocol.load_module(
       module_name="flexStackerModuleV1",
       location="C4"
   )

> In this example, Stacker shuttles occupy deck slots A3 and C3. 

Reviewer: ```suggestion
In this example, Stacker shuttles occupy deck slots A4 and C4. 
```
Maybe we should show loading one with waste chute too. I'm not sure how it's different.

Author: Looking back at the manual, looks like you just use the Waste Chute adapter rather than the regular adapter?

   )
   stacker_2.set_stored_labware(
       load_name="opentrons_96_wellplate_200ul_pcr_full_skirt",
       count=12
   )

In this example, ``stacker_1`` is configured to hold 6 Flex tip racks, each with a compatible lid. Flex tip racks must have lids to be properly stored in the Stacker. ``stacker_2`` is configured to hold 12 PCR plates without lids. You must configure each Stacker in your protocol before using :py:meth:`.FlexStackerContext.store` or :py:meth:`.FlexStackerContext.retrieve`. 

:py:meth:`~.FlexStackerContext.set_stored_labware` assigns a labware stack to the Stacker in this deck slot. You can use the Stacker and shuttle as a normal deck slot earlier in your protocol with :py:meth:`~.FlexStackerContext.load_labware`. You'll need to move this labware elsewhere on the deck before configuring and using the Stacker for storage.  

Stacker Capacity 
-----------------

Different labware have varying `z` heights. One well plate might be 2 mm taller than another, affecting the number of plates the Stacker can store at once. The API includes helper commands to calculate how many labware the Stacker can hold: 

    - `.ModuleContext.get_max_storable_labware`

>     - `.ModuleContext.get_current_storable_labware` 

Reviewer: ```suggestion
    - :py:meth:`.FlexStackerContext.get_max_storable_labware`
    - :py:meth:`.FlexStackerContext.get_current_storable_labware` 
```

    
Use ``get_max_storable_labware()`` or ``get_current_storable_labware()`` to calculate the maximum or current number of labware the Stacker can store, based on either the labware definition or the Stacker's current storage conditions.

Like other hardware modules, the Stacker doesn't verify the labware you place inside it. If your Stacker is configured correctly, you can use ``get_stored_labware()`` throughout your protocol to check an updated list of labware stored inside. 

Using Stacker Labware
======================
During a protocol, use :py:meth:`~.FlexStackerContext.retrieve` to automatically access a single piece of labware to use on the Flex deck. 

.. code-block:: python

   stacker_1.retrieve()
   protocol.move_labware(

>        labware="opentrons_flex_96_tiprack_200ul",

Reviewer: This won't work. `labware` has to be a labware object, not a string.

However, this clever bit of code _does_ work! It takes whatever the stacker retrieves and moves it.

```
protocol.move_labware(
    labware=stacker_1.retrieve(),
    new_location="B2",
    use_gripper="True"
)
```

       new_location="B2",

>        use_gripper="True"

Reviewer: this also can't be a string.
```suggestion
       use_gripper=True
```

   )

Here, the Stacker's shuttle takes the Flex tip rack at the bottom of the labware stack and moves it into slot A3. Then, the Flex Gripper moves that tip rack to slot B2 on the deck. 

> To move labware from the deck into a Stacker, use `~.FlexStackerContext.store()`::

Reviewer: ```suggestion
To move labware from the deck into a Stacker, use :py:meth:`~.FlexStackerContext.store()`::
```

   protocol.move_labware(
       labware="opentrons_96_wellplate_200ul_pcr_full_skirt",
       new_location=stacker_2,
       use_gripper="True"
   )

>    stacker_2.store()

Reviewer: ```suggestion
       labware=plate,
       new_location=stacker_2,
       use_gripper=True
   )
   stacker_2.store()
```

we tend to define `plate` as in our templates, so that's safe to use as an example.

:og:description: How to use the Flex Stacker Module in a Python protocol. 

.. _stacker:

*******************
Flex Stacker Module
*******************

> The Flex Stacker is an external module that provides automated storage of well plates, Flex tip racks, or reservoirs. The Stacker's attached shuttle moves labware from the stack to the deck for use during a protocol. 

Reviewer 2: Should the intro sentence link to the other Stacker doc?

https://docs.opentrons.com/stacker-manual/

Author: I'm not sure. I don't think we've cross referenced like this before, and our new search feature would capture both kinds of docs.

Reviewer: Probably not in this top-level context. There might be a couple other places we'd want to xref (like to the list of supported labware) but I think we should hold off until both are in mkdocs.

> The Flex Stacker is an external module that provides automated storage of well plates, Flex tip racks, or reservoirs. The Stacker's attached shuttle moves labware from the stack to the deck for use during a protocol. 

Reviewer 3: 1. Provides automated dispensal/storage, we want to enphesize dispensal.
2. Lets specify Flex tip racks, well plates, or reservoirs (Flex tipracks are the primary labware)

> The Stacker is represented in code by a ``StackerContext`` object that includes methods for storing and retrieving labware. You can also use helper commands in your protocol to calculate how many labware the Stacker can store at once. 

Reviewer 3: lets emphasize retrieving first

Loading and Deck Slots
========================

> Up to four Stacker Modules can be attached to the right side of your Flex, creating deck slots A4--D4. Each Stacker occupies a deck slot in column 4, with the shuttle in column 3. 

Reviewer 2: ```suggestion
Up to four Stacker Modules can be attached to the right side of your Flex, creating deck slots A4–D4. Each Stacker occupies a deck slot in column 4, with the shuttle in column 3. 
```

Start by loading each Stacker as you would any other module: 

.. code-block:: python

   stacker_1 = protocol.load_module(
       module_name="flexStackerModuleV1",
       location="A4"
   )
   stacker_2 = protocol.load_module(
       module_name="flexStackerModuleV1",
       location="C4"
   )

> Each Stacker occupies a deck slot in column 4, with the attached shuttle in column 3. In this example, Stacker shuttles occupy deck slots A3 and C3. 

Reviewer: Duplicates sentence above.
```suggestion
In this example, Stacker shuttles occupy deck slots A3 and C3. 
```

Adding Labware to the Stacker
==============================

> Next, you'll need to define the *type* and amount of labware the Stacker will store. Throughout yor protocol, the Flex automatically moves labware, like well plates or tip racks, from inside the Stacker to the deck. 

Reviewer: Emphasis not needed.

Each Stacker can hold a labware stack of up to:

> - 7 Flex tipracks (6 in the Stacker and 1 on the shuttle)

Reviewer: I would add "with lids" here, and keep the "tip racks must have lids" language below.

- 48 PCR plates, like <tested load_name>

> - 16 deep well plates, like <tested load_name>

Reviewer: Fill these load names in. Use Opentrons Tough labware if possible.

Author: Updated these and tested to make sure they work- Tough well plate and NEST reservoir (no tough reservoir in the labware library, unless one's being released I'm forgetting about?). 

@vegano1 can I get a triple check that these labware are OK to recommend with the Stacker?

> You'll need to use :py:meth:`~.FlexStackerContext.set_stored_labware` to configure the Stacker before adding or removing labware during a protocol. Only one type of labware can be stored in each Stacker at one time. 

Reviewer: I would move "Only one type…" to the end of the first ¶ in this section.

.. code-block:: python

   stacker_1.set_stored_labware(
       load_name="opentrons_flex_96_tiprack_200ul",
       count=6,
       lid="opentrons_flex_tiprack_lid"
   )
   stacker_2.set_stored_labware(
       load_name="opentrons_96_wellplate_200ul_pcr_full_skirt",
       count=12
   )

> In this example, `stacker_1` is configured to hold 6 Flex tip racks, each with a compatible lid. Flex tip racks must have lids to be properly stored in the Stacker. `stacker_2` is configured to hold 12 PCR plates without lids. 

Reviewer: RST 🤪 
```suggestion
In this example, ``stacker_1`` is configured to hold 6 Flex tip racks, each with a compatible lid. Flex tip racks must have lids to be properly stored in the Stacker. ``stacker_2`` is configured to hold 12 PCR plates without lids. 
```

Author: gahhh  @ markdown cooking my brain

> Configuring the Stacker assigns a labware stack to this deck slot. You can use the Stacker and shuttle as a normal deck slot earlier in your protocol with :py:meth:`~.ProtocolContext.load_labware()`. You'll need to move this labware elsewhere on the deck before configuring and using the Stacker for storage.  

Reviewer: Is this right? `ProtocolContext.load_labware()`, not `FlexStackerContext.load_labware()`?

If the latter, we'll need to unhide `FlexStackerContext.load_labware()`. And since it's an inherited method, we'll need to add a note about how it differs from loading onto other modules.

Author: oooh good call out. will check with Brayan on this one.

Author: It  should be `FlexStackerContext.load_labware()`. updating...

> .. note:: 

Reviewer: I think putting this all in a note minimizes it too much. This could be a section named "Stacker Capacity" or similar.

    Different labware have varying `z` heights. One well plate might be 2 mm taller than another, affecting the number of plates the Stacker can store at once. The API includes helper commands to calculate how many labware the Stacker can hold: 

    - `.ModuleContext.get_max_storable_labware()`
    - `.ModuleContext.get_current_storable_labware()` 

>     - `.ModuleContext.get_stored_labware()` 

Reviewer: `get_stored_labware()` seems the odd one out here. It doesn't "calculate how many labware the Stacker can hold"

I think it's best to list it later.
```suggestion
    - `.FlexStackerContext.get_max_storable_labware()`
    - `.FlexStackerContext.get_current_storable_labware()` 
```

>     Use ``get_max_storable_labware()`` or ``get_current_storable_labware()`` to calculate the maximum or current number of labware the Stacker can store, based on either the labware definition or the Stacker's current storage conditions.

Reviewer 2: Is there an indent here that shouldn't be?

Author: probably. moving this whole section from the note anyway!

>     Like other hardware modules, the Stacker doesn't verify the labware you place inside it. If your Stacker is configured correctly, you can use ``get_stored_labware()`` throughout your protocol to check an updated list of labware stored inside. 

Reviewer 2: Same as above, re: indent?

>     Like other hardware modules, the Stacker doesn't verify the labware you place inside it. If your Stacker is configured correctly, you can use ``get_stored_labware()`` throughout your protocol to check an updated list of labware stored inside. 

Reviewer: What does "doesn't verify" mean? The same as that we don't have a way to confirm that you _actually_ put a well plate directly on the deck, as the protocol said? Or that the Stacker methods don't prevent you from doing impossible things, like setting 20 tip racks as stored in the Stacker? (I would hope the latter raises an analysis error.)

Author: Yes to the first: "we don't have a way to confirm that you _actually_ put a well plate directly on the deck". Based on testing, the API does raise an error if you try to add more tip racks (or other labware) than are allowed.

Using Stacker Labware
======================

> During a protocol, use `~.FlexStackerContext.retrieve()` to automatically access a single piece of labware to use on the Flex deck. 

Reviewer: Broken link
```suggestion
During a protocol, use :py:meth:`~.FlexStackerContext.retrieve()` to automatically access a single piece of labware to use on the Flex deck. 
```

.. code-block:: python

   stacker_1.retrieve()
   protocol.move_labware(
       labware="opentrons_flex_96_tiprack_200ul",
       new_location="B2",
       use_gripper="True"
   )

> Here, the Flex tip rack at the bottom of the labware stack is moved to the shuttle in slot A3. Then, use the Flex Gripper or manually move the new tip rack elsewhere on the deck. 

Reviewer: Suggested rewording to line up the tenses and make it clear what's moving what.
```suggestion
Here, the Stacker's shuttle takes the Flex tip rack at the bottom of the labware stack and moves it into slot A3. Then, the Flex Gripper moves that tip rack to slot B2 on the deck. 
```

> To add more labware to a Stacker, use `~.FlexStackerContext.store()`::

Reviewer: ```suggestion
To move labware from the deck into a Stacker, use `~.FlexStackerContext.store()`::
```
I think the phrase "add more labware to a Stacker" more intuitively refers to manually loading it in via the external door.

Author: agreed AND this is different conditions (on deck vs off deck) for storing and filling...

   protocol.move_labware(
       labware="opentrons_96_wellplate_200ul_pcr_full_skirt",
       new_location=stacker_2,
       use_gripper="True"
   )
   stacker_2.store()

After placing labware on the shuttle in slot C3, ``stacker_2`` stores another well plate on the bottom of the stack. 

To mark a Stacker as completely full or empty, use `~.ModuleContext.stacker.fill()` or `~.StackerContext.stacker.empty()`::

   # mark the second Stacker as empty
   stacker_2.empty()

   # configure the second Stacker and fill with new labware
   stacker_2.set_stored_labware(
       load_name="nest_12_reservoir_15ml",
       count=8
   )
   stacker_2.fill(8, "Add 8 reservoirs to stacker_2.")

> Both ``fill()`` and ``empty()`` pause the protocol and allow you to manually add or remove labware from the Stacker. Each method assumes labware is loaded or moved ``off_deck``. 

Author: Calling this out for a triple check on assuming labware is loaded or moved ``off_deck``. If not true, need to remove this from API reference as well (I added it).

Reviewer: I think this detail can be omitted.

We should, however, differentiate between fill with and without the `count` argument. And make it really clear that you have to specify the type in `set_stored_labware`. (It seems very unintuitive to me that `fill` doesn't take a load name argument. But that's how it is.)

Author: Updated in a few places to make this clearer.

> 

Reviewer: Please zap these extra newlines.
